plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+	for _, serviceAccountName := range set.ServiceAccountNames {
+		// checkOut is its own function to avoid deferring the lock release inside a loop,
+		// which would cause a memory leak.
+		password, err := b.checkOut(ctx, req.Storage, serviceAccountName, newCheckOut)
+		if err == ErrCheckedOut {
+			continue
+		}
+		respData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+			"password":             password,
+		}
+		internalData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+		}
+		resp := b.Backend.Secret(secretAccessKeyType).Response(respData, internalData)
+		resp.Secret.Renewable = true
+		resp.Secret.TTL = ttl
+		resp.Secret.MaxTTL = set.MaxTTL
+		return resp, nil
+	}
+
+	// If we arrived here, it's because we never had a hit for a service account that was available.
+	resp, err := logical.RespondWithStatusCode(&logical.Response{
+		Warnings: []string{"No service accounts available for check-out, please try again later."},
+	}, req, 429)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (b *backend) checkOut(ctx context.Context, storage logical.Storage, serviceAccountName string, newCheckOut *CheckOut) (string, error) {
+	lock := locksutil.LockForKey(b.checkOutLocks, serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	// We've found a service account that we can check out. Let's do it and respond.
+	if err := b.checkOutHandler.CheckOut(ctx, storage, serviceAccountName, newCheckOut); err != nil {
+		return "", err
+	}
+	password, err := retrievePassword(ctx, storage, serviceAccountName)
+	if err != nil {
+		return "", err
+	}
+	return password, nil
+}
+
+func (b *backend) secretAccessKeys() *framework.Secret {
+	return &framework.Secret{
+		Type: secretAccessKeyType,
+		Fields: map[string]*framework.FieldSchema{
+			"service_account_name": {
+				Type:        framework.TypeString,
+				Description: "Access Key",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Secret Key",
+			},
+		},
+		Renew:  b.renewCheckOut,
+		Revoke: b.endCheckOut,
+	}
+}
+
+func (b *backend) renewCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	serviceAccountName := req.Secret.InternalData["service_account_name"].(string)
+
+	lock := locksutil.LockForKey(b.checkOutLocks, serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, serviceAccountName)
+	if err != nil {
+		return nil, err
+	}
+	if checkOut.IsAvailable {
+		// It's possible that this renewal could be attempted after a check-in occurred either by this entity or by
+		// another user with access to the "manage check-ins" endpoint that forcibly checked it back in.
+		return logical.ErrorResponse(fmt.Sprintf("%s is already checked in, please call check-out to regain it", serviceAccountName)), nil
+	}
+	return &logical.Response{Secret: req.Secret}, nil
```

Comment:
I think we'd want to update the TTL and MaxTTL with their current values at the time of renewal.

Reply:
I tested that and actually, we don't. I saw that in the AWS secrets engine. I found that if I let the leasing functionality handle it, it does the right thing without further manipulation.

Reply:
Marking this one as resolved because it's behaving the way we want currently.

Reply:
How does the leasing system know if the set config has been changed to new TTLs though? The information is in the set storage. From what I've seen the usual pattern is to reread the role/set on rename and update the TTLs so I'm curious how this is currently working correctly.

Reply:
Ah, I see! Good point. I tried that out and tested it through and it behaves in the intended way. I'll push that update.